stream_alert/rule_processor/sink.py

```diff
@@ -92,3 +95,88 @@ def sink(self, alerts):
                             self.function,
                             response['ResponseMetadata']['RequestId'])
+
+    def _alert_batches(self, alerts, batch_size=25):
+        """Group alerts into batches of 25, the maximum allowed by Dynamo batch_write_item.
+
+        Yields:
+            (dict) The constructed request for batch_write_item, containing <= 25 alerts.
+                Maps table name to a list of requests.
+        """
+        for i in xrange(0, len(alerts), batch_size):
+            batch = alerts[i:i+batch_size]
+            yield {
+                self.table: [
+                    {
+                        'PutRequest': {
+                            'Item': {
+                                'RuleName': {'S': alert['rule_name']},
+                                'Timestamp': {
+                                    # ISO 8601 datetime format, and is unique for each alert
+                                    'S': datetime.now().strftime('%Y-%m-%dT%H:%M:%S.%fZ')
+                                },
+                                'Cluster': {'S': os.environ['CLUSTER']},
+                                'RuleDescription': {'S': alert['rule_description']},
+                                'Outputs': {'SS': alert['outputs']},
+                                # Compact JSON encoding (no extra spaces)
+                                'Record': {'S': json.dumps(alert['record'], separators=(',', ':'))}
+                            }
+                        }
+                    }
+                    for alert in batch
+                ]
+            }
+
+    @backoff.on_exception(backoff.expo, ClientError, max_tries=5, jitter=backoff.full_jitter,
+                          on_backoff=backoff_handlers.backoff_handler,
+                          on_success=backoff_handlers.success_handler,
+                          on_giveup=backoff_handlers.giveup_handler)
+    def _batch_write(self, request_items, max_attempts=5):
+        """Write a batch of alerts to Dynamo, retrying with exponential backoff for failed items.
+
+        Args:
+            request_items (dict): The generated RequestItems dict from _alert_batches().
+            max_attempts (int): Maximum number of times to retry UnprocessedItems.
+
+        Returns:
+            (bool) True if the batch write was eventually successful, False otherwise.
+        """
+        for attempt in xrange(1, max_attempts + 1):
+            response = self.client_dynamo.batch_write_item(RequestItems=request_items)
+
+            # If Dynamo experiences an internal error, unprocessed items are listed in the response.
+            # AWS recommends retrying unprocessed items in a loop with exponential backoff.
+            request_items = response['UnprocessedItems']
+            if request_items:
+                LOGGER.warn(
+                    'Batch write failed: %d alerts were not written (attempt %d/%d)',
```

Comment:
Do you think we should add a custom metric for this right away so we have some insight during the testing phase?

Reply:
Nah - there are no metrics in the alert processor right now, and adding support for them would be nontrivial. We can just search `Batch write failed` in the logs - it shouldn't ever happen during normal operation.

But I'm amenable if you feel strongly we should add metrics sooner than later. I think as the alert merger develops, then it may make more sense

Reply:
@austinbyers This code is in the rule processor :)

Reply:
This has now been replaced with a @backoff decorator, which has its own logging which we can search for